### ISSUE TITLE
gdbserver reverse continue/step support

### ISSUE THREAD

**Issue by Manouchehri** [radareorg/radare2#8966]
```
vagrant@vagrant:~$ r2 -D gdb gdb://127.0.0.1:4444 `which curl`
= attach 6371 1
= attach 6371 0
 -- Get a free shell with 'ragg2 -i exec -x'
[0x00006dc0]> dc
= attach 6371 0
[+] signal 9 aka SIGKILL received 0
[0x70000002]> dcb
cannot continue back
[0x70000002]>
```
```
vagrant@vagrant:~$ r2 -D gdb gdb://127.0.0.1:4444 `which curl`
= attach 6371 1
= attach 6371 0
 -- Run .dmm* to load the flags of the symbols of all modules loaded in the debugger
[0x00006dc0]> ds 16
= attach 6371 6371
[0x7f4c386849d6]> dsb
cannot step back
```

Example session with gdb:

```
vagrant@vagrant:~$ gdb -q `which curl`
Reading symbols from /usr/bin/curl...(no debugging symbols found)...done.
(gdb) target remote localhost:4444
Remote debugging using localhost:4444
0x00007f4c38683c30 in ?? ()
(gdb) c
Continuing.

Program received signal SIGKILL, Killed.
0x0000000070000002 in ?? ()
(gdb) rc
Continuing.

Program stopped.
0x00007f4c38683c30 in ?? ()
(gdb) c
Continuing.

Program received signal SIGKILL, Killed.
0x0000000070000002 in ?? ()
(gdb) rsi 1
0x00007f4c3845a3ed in ?? ()
(gdb) si
0x0000000070000000 in ?? ()
(gdb) rsi 16
0x00007f4c38456be3 in ?? ()
```

**Comment by radare** [comment]
The gdb thing doesnt works like that but the crash must be fixed

**Comment by XVilka** [comment]
@radare no, it's just a matter of supporting reverse debugging in gdb protocol, and connecting it with reverse step/continue commands.

**Comment by yossizap** [comment]
>no, it's just a matter of supporting reverse debugging in gdb protocol, and connecting it with reverse step/continue commands

@XVilka why should r2 depend on gdbserver's `record` implementation, which isn't always available, when it's possible to simply set the state from r2's history as in any other debugger?

It seems that a wrong implementation/usage of gdbserver's '[G](https://sourceware.org/gdb/onlinedocs/gdb/Packets.html#write-registers-packet)' (write registers) is the issue. Currently looking into this.

For reference, gdbserver's output after r_debug_session_set:
`gdbserver: Wrong sized register packet (expected 1792 bytes, got 4)`
Printed by [registers_from_string](https://github.com/bminor/binutils-gdb/blob/master/gdb/gdbserver/regcache.c#L233) which is called by [gdb/gdbserver/server.c](https://github.com/bminor/binutils-gdb/blob/master/gdb/gdbserver/server.c#L4132)'s command parsing loop.

**Comment by XVilka** [comment]
@yossizap it's not a requirement. Whatever works best. Thanks for working on this! 💯

**Comment by yossizap** [comment]
Solved the reg writing protocol issue in #15455 but it still doesn't work since [`r_reg_next_diff`](https://github.com/radareorg/radare2/blob/master/libr/reg/reg.c#L353) which is iterated on by [`r_debug_write_reg`](https://github.com/radareorg/radare2/blob/master/libr/debug/p/debug_gdb.c#L273) only returns rax and orig_rax, other registers can't be found in [`r_reg_get_at`](https://github.com/radareorg/radare2/blob/master/libr/reg/reg.c#L337) due to size and offset mismatches. 
Tried simply writing `buf` with [`gdbr_write_bin_registers`](https://github.com/radareorg/radare2/blob/master/shlr/gdb/src/gdbclient/core.c#L959)(changed args to receive buf and len instead of taking it from `g->data`), as other debuggers do, but it wrote lots of wrong values instead. This seems very weird to me since afaik, that buf should be the same buf that read_regs returned, which is exactly how [`gdbr_write_reg` does it](https://github.com/radareorg/radare2/blob/master/shlr/gdb/src/gdbclient/core.c#L1070).

I am probably missing something so I would appreciate some clues. Also seems that gdbr and qnx are the only ones using `r_reg_next_diff`, could be outdated? `dr` doesn't seem to have an issue finding the delta between commands but it compares two reg lists while dreg.c still uses bufs.

**Comment by yossizap** [comment]
Solved the previous issue but `dsb` is partially broken and `dcb` has issues when stepping back multiple times in a row, could be a limit on session size?

**Comment by radare** [comment]
fixed in master

**Comment by XVilka** [comment]
Not completely - see the PR description, it supports only RR, but not the rest.

**Comment by yossizap** [comment]
*RR and any other gdbserver with ReverseContinue and ReverseStep, could also work with r2 gdbserver etc.

For proper reverse debug with regular gdbserver we'll need to do it on r2's side by fixing dsb/dcb.
